build error: no matching function for call to ‘push_back(digidoc::Container*...

I am having some trouble building from master and I am not sure where this one comes from?

Can someone help?

```sh
[  1%] Built target docs
[ 40%] Built target digidocpp_priv
[ 45%] Built target minizip
[ 46%] Building CXX object src/CMakeFiles/digidocpp.dir/Container.cpp.o
/home/hector/git/libdigidocpp/src/Container.cpp: In instantiation of ‘static void digidoc::Container::addContainerImplementation() [with T = digidoc::PDF]’:
/home/hector/git/libdigidocpp/src/Container.cpp:158:48:   required from here
/home/hector/git/libdigidocpp/src/Container.cpp:423:27: error: no matching function for call to ‘push_back(digidoc::Container* (*)(const string&))’
  423 |     m_createList.push_back(&T::createInternal);
      |     ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/10/vector:67,
                 from /home/hector/git/libdigidocpp/src/Container.h:26,
                 from /home/hector/git/libdigidocpp/src/Container.cpp:20:
/usr/include/c++/10/bits/stl_vector.h:1187:7: note: candidate: ‘void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&); _Alloc = std::allocator<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>; std::vector<_Tp, _Alloc>::value_type = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)]’ (near match)
 1187 |       push_back(const value_type& __x)
      |       ^~~~~~~~~
/usr/include/c++/10/bits/stl_vector.h:1187:7: note:   conversion of argument 1 would be ill-formed:
/home/hector/git/libdigidocpp/src/Container.cpp:423:27: error: invalid conversion from ‘digidoc::Container* (*)(const string&)’ {aka ‘digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)’} to ‘std::vector<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>::value_type’ {aka ‘std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)’} [-fpermissive]
  423 |     m_createList.push_back(&T::createInternal);
      |     ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
      |                           |
      |                           digidoc::Container* (*)(const string&) {aka digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)}
In file included from /usr/include/c++/10/vector:67,
                 from /home/hector/git/libdigidocpp/src/Container.h:26,
                 from /home/hector/git/libdigidocpp/src/Container.cpp:20:
/usr/include/c++/10/bits/stl_vector.h:1203:7: note: candidate: ‘void std::vector<_Tp, _Alloc>::push_back(std::vector<_Tp, _Alloc>::value_type&&) [with _Tp = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&); _Alloc = std::allocator<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>; std::vector<_Tp, _Alloc>::value_type = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)]’ (near match)
 1203 |       push_back(value_type&& __x)
      |       ^~~~~~~~~
/usr/include/c++/10/bits/stl_vector.h:1203:7: note:   conversion of argument 1 would be ill-formed:
/home/hector/git/libdigidocpp/src/Container.cpp:423:27: error: invalid conversion from ‘digidoc::Container* (*)(const string&)’ {aka ‘digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)’} to ‘std::vector<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>::value_type’ {aka ‘std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)’} [-fpermissive]
  423 |     m_createList.push_back(&T::createInternal);
      |     ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
      |                           |
      |                           digidoc::Container* (*)(const string&) {aka digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)}
/home/hector/git/libdigidocpp/src/Container.cpp:424:25: error: no matching function for call to ‘push_back(digidoc::Container* (*)(const string&))’
  424 |     m_openList.push_back(&T::openInternal);
      |     ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/10/vector:67,
                 from /home/hector/git/libdigidocpp/src/Container.h:26,
                 from /home/hector/git/libdigidocpp/src/Container.cpp:20:
/usr/include/c++/10/bits/stl_vector.h:1187:7: note: candidate: ‘void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&); _Alloc = std::allocator<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>; std::vector<_Tp, _Alloc>::value_type = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)]’ (near match)
 1187 |       push_back(const value_type& __x)
      |       ^~~~~~~~~
/usr/include/c++/10/bits/stl_vector.h:1187:7: note:   conversion of argument 1 would be ill-formed:
/home/hector/git/libdigidocpp/src/Container.cpp:424:25: error: invalid conversion from ‘digidoc::Container* (*)(const string&)’ {aka ‘digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)’} to ‘std::vector<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>::value_type’ {aka ‘std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)’} [-fpermissive]
  424 |     m_openList.push_back(&T::openInternal);
      |     ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
      |                         |
      |                         digidoc::Container* (*)(const string&) {aka digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)}
In file included from /usr/include/c++/10/vector:67,
                 from /home/hector/git/libdigidocpp/src/Container.h:26,
                 from /home/hector/git/libdigidocpp/src/Container.cpp:20:
/usr/include/c++/10/bits/stl_vector.h:1203:7: note: candidate: ‘void std::vector<_Tp, _Alloc>::push_back(std::vector<_Tp, _Alloc>::value_type&&) [with _Tp = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&); _Alloc = std::allocator<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>; std::vector<_Tp, _Alloc>::value_type = std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)]’ (near match)
 1203 |       push_back(value_type&& __x)
      |       ^~~~~~~~~
/usr/include/c++/10/bits/stl_vector.h:1203:7: note:   conversion of argument 1 would be ill-formed:
/home/hector/git/libdigidocpp/src/Container.cpp:424:25: error: invalid conversion from ‘digidoc::Container* (*)(const string&)’ {aka ‘digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)’} to ‘std::vector<std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)>::value_type’ {aka ‘std::unique_ptr<digidoc::Container> (*)(const std::__cxx11::basic_string<char>&)’} [-fpermissive]
  424 |     m_openList.push_back(&T::openInternal);
      |     ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
      |                         |
      |                         digidoc::Container* (*)(const string&) {aka digidoc::Container* (*)(const std::__cxx11::basic_string<char>&)}
make[2]: *** [src/CMakeFiles/digidocpp.dir/build.make:194: src/CMakeFiles/digidocpp.dir/Container.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:222: src/CMakeFiles/digidocpp.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

--- a/src/PDF.cpp
+++ b/src/PDF.cpp
@@ -180,7 +180,7 @@ void PDF::addAdESSignature(istream & /*signature*/)
     THROW("Not supported.");
 }
 
-Container* PDF::createInternal(const string & /*unused*/)
+unique_ptr<Container> PDF::createInternal(const string & /*path*/)
 {
     return nullptr;
 }
@@ -195,22 +195,19 @@ string PDF::mediaType() const
     return "application/pdf";
 }
 
-Container* PDF::openInternal(const string &path)
+unique_ptr<Container> PDF::openInternal(const string &path)
 {
     if(File::fileExtension(path) != "pdf")
-        return nullptr;
+        return {};
 
-    DEBUG("DDoc::openInternal(%s)", path.c_str());
-    ifstream *is = new ifstream(File::encodeName(path).c_str(), ifstream::binary);
+    DEBUG("PDF:openInternal(%s)", path.c_str());
+    unique_ptr<ifstream> is(new ifstream(File::encodeName(path).c_str(), ifstream::binary));
     string line;
     getline(*is, line);
     if(line.compare(0, 7, "%PDF-1.") != 0)
-    {
-        delete is;
-        return nullptr;
-    }
+        return {};
 
-    PDF *doc = new PDF(path);
+    unique_ptr<PDF> doc(new PDF(path));
     try {
         PdfMemDocument parser(path.c_str());
         for(const PdfObject *obj: parser.GetObjects())
@@ -279,12 +276,10 @@ Container* PDF::openInternal(const string &path)
             }
         }
     } catch(const PdfError &e) {
-        delete is;
-        delete doc;
         THROW_CAUSE(EXCEPTION(e.what()), "Failed to parse PDF.");
     }
 
-    doc->d->dataFiles.push_back(new DataFilePrivate(is, File::fileName(path), "application/pdf", File::fileName(path)));
+    doc->d->dataFiles.push_back(new DataFilePrivate(move(is), File::fileName(path), "application/pdf", File::fileName(path)));
     return doc;
 }
 

--- a/src/PDF.h
+++ b/src/PDF.h
@@ -27,16 +27,15 @@
 namespace digidoc
 {
 
-class SignaturePDF;
 class PDF: public Container
 {
 public:
     ~PDF();
-    void save(const std::string &path = "") override;
+    void save(const std::string &path = {}) override;
     std::string mediaType() const override;
 
     void addDataFile(const std::string &path, const std::string &mediaType) override;
-    void addDataFile(std::istream *is, const std::string &fileName, const std::string &mediaType) override;
+    void addDataFile(std::unique_ptr<std::istream> is, const std::string &fileName, const std::string &mediaType) override;
     std::vector<DataFile*> dataFiles() const override;
     void removeDataFile(unsigned int id) override;
 
@@ -46,8 +45,8 @@ public:
     void removeSignature(unsigned int id) override;
     Signature* sign(Signer* signer) override;
 
-    static Container* createInternal(const std::string &path);
-    static Container* openInternal(const std::string &path);
+    static std::unique_ptr<Container> createInternal(const std::string &path);
+    static std::unique_ptr<Container> openInternal(const std::string &path);
 
 private:
     PDF(const std::string &path);

